storage: Don't try to replicate ranges that need to split

bdarnell opened the pull request:
Fixes #6444
Fixes half of #6400

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/6447)

<!-- Reviewable:end -->

bdarnell committed:
storage: Don't try to replicate ranges that need to split

Fixes #6444
Fixes half of #6400

tschottdorf commented on storage/replica.go:
```diff
@@ -2111,6 +2111,13 @@ func (r *Replica) loadSystemConfigSpan() ([]roachpb.KeyValue, []byte, error) {
+// to be split.
+func (r *Replica) needsSplitBySize() bool {
+	maxBytes := r.GetMaxBytes()
+	return maxBytes > 0 && r.stats.KeyBytes+r.stats.ValBytes > maxBytes
```
how is accessing `r.stats` not racy?

tamird replied:
It is, that's what failed in CI.

On Mon, May 2, 2016 at 4:11 PM, Tobias Schottdorf <notifications@github.com>
wrote:

> In storage/replica.go
> <https://github.com/cockroachdb/cockroach/pull/6447#discussion_r61795681>:
>
> > @@ -2111,6 +2111,13 @@ func (r *Replica) loadSystemConfigSpan() ([]roachpb.KeyValue, []byte, error) {
> >  	return kvs, config.SystemConfig{Values: kvs}.Hash(), nil
> >  }
> >
> > +// needsSplitBySize returns true if the size of the range requires it
> > +// to be split.
> > +func (r *Replica) needsSplitBySize() bool {
> > +	maxBytes := r.GetMaxBytes()
> > +	return maxBytes > 0 && r.stats.KeyBytes+r.stats.ValBytes > maxBytes
>
> how is accessing r.stats not racy?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/6447/files/89b63187315656c641c369176a87e427f22a5f32#r61795681>
>

tschottdorf replied:
I see. Quick grep for complete list of racy candidates (on master)

```
storage/gc_queue.go|129 col 26| gcScore := float64(repl.stats.GetGCBytesAge(now.WallTime)) / float64(zone.GC.TTLSeconds) / float64(gcByteCountNormalization)
storage/gc_queue.go|133 col 22| intentScore := repl.stats.GetAvgIntentAge(now.WallTime) / float64(intentAgeNormalization.Nanoseconds()/1E9)
storage/replica.go|2119 col 23| if maxBytes > 0 && r.stats.KeyBytes+r.stats.ValBytes > maxBytes {
storage/replica.go|2119 col 40| if maxBytes > 0 && r.stats.KeyBytes+r.stats.ValBytes > maxBytes {
```

bdarnell replied:
Hmm, how was it working in `maybeAddToSplitQueue` then? (and I see I only did half the job of extracting this code from that method) I guess it was protected by virtue of being in the processRaft goroutine? Synchronized this correctly and un-embedded `rangeStats.MVCCStats` to discourage direct unsynchronized access to members.

tschottdorf commented:
LGTM mod races

bdarnell committed:
storage: Make rangeStats.mvccStats a named field

Embedding this struct invites direct unsynchronized access to its
internal state; the getter methods should be used instead.

tschottdorf commented on storage/stats.go:
```diff
@@ -34,9 +34,9 @@ import (
-	engine.MVCCStats // embedded, cached version of stat values
+	rangeID    roachpb.RangeID
+	sync.Mutex                  // Protects mvccStats
+	mvccStats  engine.MVCCStats // embedded, cached version of stat values
```
s/embedded, //

bdarnell replied:
Done, and updated some other outdated comments in this file.

tschottdorf commented:
LGTM
